### ISSUE TITLE
Inconsistent treatment of single statement lambda vs. block

### ISSUE THREAD

**Issue by olafurpg** [scalameta/scalafmt#1826]
- **Version**: 2.4.2
- **Integration**: cli
- **Configuration**:  defaults

## Steps

Given code like this:
```scala
files.test("basic") { name =>
  assertEquals(name, "basic-setup")
}
files.test("basic") {
  assertEquals(name, "basic-setup")
}
```

## Problem

Scalafmt formats code like this:
```scala
files.test("basic") { name => assertEquals(name, "basic-setup") }
files.test("basic") {
  assertEquals(name, "basic-setup")
}
```

## Expectation

I would like the formatted output to ideally look unchanged.

```scala
files.test("basic") { name =>
  assertEquals(name, "basic-setup")
}
files.test("basic") {
  assertEquals(name, "basic-setup")
}
```

Alternatively, the block should be treated consistently with the lambda and look like this
```scala
files.test("basic") { name => assertEquals(name, "basic-setup") }
files.test("basic") { assertEquals(name, "basic-setup") }
```

## Notes

This issue is specifically about the default behavior. I'm OK with keeping the current behavior under a non-default setting.

**Comment by poslegm** [comment]
I think the correct solution is to hide it under `newlines.source=fold` like in the https://github.com/scalameta/scalafmt/issues/1809

**Comment by kitbellew** [comment]
@olafurpg part of the issue with consistency is that lambda is guided by `newlines.afterCurlyLambda` parameter.

however, since you asked for default behaviour, in which this parameter is set to `never`, i think the solution is to fold the non-lambda block when it's an apply argument, and not in other cases.

thoughts? + @poslegm 

p.s. `newlines.source=fold` should already cover this case, but removing the lambda behaviour might upset those who submitted the issue fixed in #1660 and #1663.

**Comment by poslegm** [comment]
> p.s. newlines.source=fold should already cover this case, but removing the lambda behaviour might upset those who submitted the issue fixed in #1660 and #1663.

Squashing lambdas only with `newlines.source=fold` looks like the most consistent and convenient solution. Users who want such behavior can just enable this config

**Comment by kitbellew** [comment]
sadly, i am expecting a loud howl if we undo #1663 which has already been released.

also, while i hope that `newlines.source` catches on, it starts out as beta and we need to wait until everyone had a chance to try it out and comment.

(i am hoping to release `newlines.source` soon but have spent the last 10 days working on the last newline-sensitive piece, handling infixes, and it's non-trivial; i wonder if @olafurpg can share why he gave up on infix handling earlier.)

**Comment by poslegm** [comment]
> sadly, i am expecting a loud howl if we undo #1663 which has already been released.

I think I missed something: why `RedundantBraces` from #1663 revert is necessary to undo
```scala
<<<
files.test("basic") { name =>
  assertEquals(name, "basic-setup")
}
>>>
files.test("basic") { name => assertEquals(name, "basic-setup") }
```
in default behavior?

**Comment by kitbellew** [comment]
the request in 1027 was to format
```
files.test("basic") { name =>
  assertEquals(name, "basic-setup")
}
```
as
```
files.test("basic")(name => assertEquals(name, "basic-setup"))
```
if it fits. and the way it is implemented is by getting this first
```
files.test("basic") { name => assertEquals(name, "basic-setup") }
```
that's why i am suggesting changing the block treatment for the apply case only.

**Comment by olafurpg** [comment]
I'm OK with having a non-default option that aggressively puts blocks/lambdas on a single line. I think the default behavior should respect the case when there is a newline in a block lambda so that multiline-block lambdas never get folded into a single line.

>  handling infixes, and it's non-trivial; i wonder if @olafurpg can share why he gave up on infix handling earlier

I gave up on infix handling since a lot of users had different preferences for how to format infix operators. I would LOVE if Scalafmt could take control over newlines in infix operators even if it's not perfect. I've seen so many cases where Scalafmt currently produces non-desirable output because the line breaks around infix operators are frozen.

**Comment by kitbellew** [comment]
> I'm OK with having a non-default option that aggressively puts blocks/lambdas on a single line. I think the default behavior should respect the case when there is a newline in a block lambda so that multiline-block lambdas never get folded into a single line.

in that case, the default for `newlines.afterCurlyLambda` should be set to `preserve` (from `never` as it is today). would that work? that's the only way to reconcile what you are asking for with the request #1027 which was solved by #1660 and #1663.

> I gave up on infix handling since a lot of users had different preferences for how to format infix operators. I would LOVE if Scalafmt could take control over newlines in infix operators even if it's not perfect. I've seen so many cases where Scalafmt currently produces non-desirable output because the line breaks around infix operators are frozen.

ok, will try some more :) generally, the code is there (and i think it does a decent job), the biggest problem is indentation (it's somehow slightly different from what it is today, although i logically understand why it comes out like that). there's also the issue of search state, but on `scala-repos` the new code fixes a few cases while breaking some others, so a wash...
